modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala

```diff
@@ -33,100 +36,164 @@ trait ConstructionService[F[_]] {
-      if (isSigned) {
-        parseSignedTransaction(hex)
-      } else {
-        parseUnsignedTransaction(hex)
+        result.leftMap(_ => MalformedTransaction)
       }
 
-    def combineTransaction(hex: Hex, signature: RosettaSignature): EitherT[F, ConstructionError, Hex] =
-      EitherT
-        .fromEither(KryoSerializer[F].deserialize[DAGTransaction](hex.toBytes))
-        .leftMap(_ => MalformedTransaction)
-        .flatMap { transaction =>
-          EitherT {
-            signature.publicKey.hexBytes.toPublicKeyByEC
-              .map(pk => SignatureProof(pk.toId, Signature(signature.hexBytes)).asRight[ConstructionError])
-          }.flatMap { proof =>
-            EitherT
-              .fromEither(
-                KryoSerializer[F]
-                  .serialize(Signed[DAGTransaction](transaction, NonEmptySet.of(proof)))
-              )
-              .leftMap(_ => MalformedTransaction.asInstanceOf[ConstructionError])
-              .map(Hex.fromBytes(_))
-          }
+      def parseUnsignedTransaction(hex: Hex): EitherT[F, ConstructionError, ConstructionParse.ParseResult] =
+        for {
+          unsignedTransaction <- KryoSerializer[F].deserialize[Transaction](hex.toBytes).toEitherT.leftMap(_ => MalformedTransaction)
+          operations <- transactionToOperations(unsignedTransaction).toEitherT
+        } yield ConstructionParse.ParseResult(operations, none)
+
+      def parseTransaction(hex: Hex, isSigned: Boolean): EitherT[F, ConstructionError, ConstructionParse.ParseResult] =
+        if (isSigned) {
+          parseSignedTransaction(hex)
+        } else {
+          parseUnsignedTransaction(hex)
         }
 
-  }
+      def combineTransaction(hex: Hex, signature: RosettaSignature): EitherT[F, ConstructionError, Hex] =
+        EitherT
+          .fromEither(KryoSerializer[F].deserialize[DAGTransaction](hex.toBytes))
+          .leftMap(_ => MalformedTransaction)
+          .flatMap { transaction =>
+            EitherT {
+              signature.publicKey.hexBytes.toPublicKeyByEC
+                .map(pk => SignatureProof(pk.toId, Signature(signature.hexBytes)).asRight[ConstructionError])
+            }.flatMap { proof =>
+              EitherT
+                .fromEither(
+                  KryoSerializer[F]
+                    .serialize(Signed[DAGTransaction](transaction, NonEmptySet.of(proof)))
+                )
+                .leftMap(_ => MalformedTransaction.asInstanceOf[ConstructionError])
+                .map(Hex.fromBytes(_))
+            }
+          }
+
+      override def getPayloads(
+        operations: NonEmptyList[Operation],
+        metadata: TransactionMetadata
+      ): EitherT[F, ConstructionError, PayloadsResult] =
+        for {
+          _ <- EitherT.cond(operations.size == 2, (), IncorrectNumberOfOperations)
+
+          (positiveOperation, negativeOperation) <- EitherT.fromEither {
+            val first = operations.head
+            val second = operations.tail.head
```

Comment:
instead of checking the size first, and then assuming that there are two operations and `head` is safe, I'd either pattern match on the operations or use headOption and then map None to IncorrectNumberOfOperations or sth like that